### PR TITLE
U/jrbogart/newer_tophat_SEDs

### PR DESCRIPTION
**Pull request by JoanneBogart**
* More accurate conversion of tophat SED to table. 
* Use sed_tools.py rather than sed_utils.py.  (Latter should be removed along with obsolete code invoking it in a future PR.) 
* Make better use of galsim for improved performance.
* Add BaseObject.write_sed( ..)
* A couple fixes for bugs which would not have been evident in typical usage

### CODE REVIEW

File: python/desc/skycatalogs/utils/sed_tools.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+import numpy as np
+import numpy.ma as ma
+from numpy.random import default_rng
+from dust_extinction.parameter_averages import F19
+import galsim
+
+__all__ = ['ObservedSedFactory', 'Extinguisher', 'AB_mag', 'get_star_sed_path']
+class ObservedSedFactory:
+    _clight = 3e8  # m/s
+    _to_W_per_Hz = 4.4659e13 # conversion of cosmoDC2 tophat Lnu values to W/Hz
```
Comment (jchiang87) [inline python/desc/skycatalogs/utils/sed_tools.py:17]: Maybe add a link to the schema in gcr-catalogs where this conversion is found: https://github.com/LSSTDESC/gcr-catalogs/blob/master/GCRCatalogs/SCHEMA.md

File: python/desc/skycatalogs/utils/sed_tools.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        sed = sed*mw_ext
+
+        return sed
+
+class AB_mag:
+    """
+    Convert flux to AB magnitude for a set of bandpasses.
+    """
+    def __init__(self, bps):
+        ab_sed = galsim.SED(lambda nu : 3631e-23, wave_type='nm',
```
Comment (jchiang87) [inline python/desc/skycatalogs/utils/sed_tools.py:184]: This is actually an approximation to the exact definition (see https://en.wikipedia.org/wiki/AB_magnitude).  Using the definition, this should be
```
ab_sed = galsim.SED(lambda nu : 10**(8.90/2.5 - 23), wave_type='nm', flux_type='fnu')
```

**Review by jchiang87 — CHANGES_REQUESTED** [review]
Looks basically fine.  I have a handful of minor change requests.

File: python/desc/skycatalogs/skyCatalogs.py [hunk 3]
```diff
@@ -196,6 +204,14 @@ def raw_config(self):
         '''
         return self._config
 
+    @property
```
Comment (jchiang87) [inline python/desc/skycatalogs/skyCatalogs.py:207]: I just noticed that `observed_sed_factory` and `extinguisher` are already defined above.

File: python/desc/skycatalogs/utils/sed_tools.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        self.wl_extinct = wl_extinct
+        self.wl_extinct_nm = wl_extinct*u.nm
+        self.extinction = F19(Rv=mwRv)
+
+        my_wl = sed_factory.wl_deltas
+        my_wl = my_wl[np.where((wl_ext_min < my_wl) & (my_wl < wl_ext_max))]
+        self.wl_deltas = my_wl
+        self.wl_deltas_u = my_wl*u.nm
+
+    def extinguish(self, sed, mwAv):
```
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:171]: Ok, but F19 does have a method `extinguish`.
Do you have a suggested replacement for `extinguisher`?

[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:171]: reddener?

File: python/desc/skycatalogs/utils/sed_tools.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+import numpy as np
+import numpy.ma as ma
+from numpy.random import default_rng
+from dust_extinction.parameter_averages import F19
+import galsim
+
+__all__ = ['ObservedSedFactory', 'Extinguisher', 'AB_mag', 'get_star_sed_path']
+class ObservedSedFactory:
+    _clight = 3e8  # m/s
```
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:16]: `astropy.constants.c` is already in units of m/s but to make it explicit one can do   `..to('m/s')` which is more compact.

File: python/desc/skycatalogs/catalog_creator.py [hunk 6]
```diff
@@ -363,10 +369,14 @@ def _sed_disk_key(s):
             knots_seds = (np.array([df[kdn] for kdn in sed_knot_names]).T).tolist()
 
         #  Compute mag_norm from TH sed and redshift
-        bulge_magnorm = [self._mag_norm_f(s[NORMWV_IX], r) for (s, r) in zip(bulge_seds, df['redshiftHubble']) ]
-        disk_magnorm = [self._mag_norm_f(s[NORMWV_IX], r) for (s, r) in zip(disk_seds, df['redshiftHubble']) ]
+        ###bulge_magnorm = [self._mag_norm_f(s[NORMWV_IX], r) for (s, r) in zip(bulge_seds, df['redshiftHubble']) ]
+        bulge_magnorm = [self._obs_sed_factory.magnorm(s, r) for (s, r)\
```
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/catalog_creator.py:373]: Fixed in next push.

File: python/desc/skycatalogs/utils/sed_tools.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:16]: Fixed in next push.

File: python/desc/skycatalogs/utils/sed_tools.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        return sed
+
+    def magnorm(self, tophat_values, z_H):
+        one_Jy = 1e-26  # W/Hz/m**2
+        Lnu = tophat_values[self.ix_500nm]*self._to_W_per_Hz  # convert to W/Hz
+        Fnu = Lnu/4/np.pi/self.dl(z_H)**2
+        with np.errstate(divide='ignore', invalid='ignore'):
+            return -2.5*np.log10(Fnu/one_Jy) + 8.90
+
+class Extinguisher:
```
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:146]: Fixed in next push

File: python/desc/skycatalogs/objects/base_object.py [hunk 8]
```diff
@@ -204,46 +214,44 @@ def get_sed(self, component=None, resolution=None):
...
+
+        f_nu500 = th_val[sky_cat.observed_sed_factory.ix_500nm]
+
+        r_h = self.get_native_attribute('redshift_hubble')
+        r = self.get_native_attribute('redshift')
+
+        sed = sky_cat.observed_sed_factory.create(th_val, r_h, r,
+                                                     resolution=resolution)
+
+        return sed, f_nu500
```
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/objects/base_object.py:244]: Fixed in next push

File: python/desc/skycatalogs/skyCatalogs.py (hunk 3 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/skyCatalogs.py:207]: Fixed in next push.

File: python/desc/skycatalogs/utils/sed_tools.py (hunk 2 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:184]: Fixed in next push

File: python/desc/skycatalogs/utils/sed_tools.py (hunk 1 above)
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:17]: Fixed in next push

File: python/desc/skycatalogs/utils/sed_tools.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoanneBogart) [reply python/desc/skycatalogs/utils/sed_tools.py:176]: I'm changing the class name as you suggested but left the rest alone for now.

**Review by jchiang87 — APPROVED** [review]
Thanks for the changes.  LGTM.

**Pull request merged**